File: commands/stats/playerstats.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone
from collections import Counter
from database import db

class PlayerStats(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='playerstats', description='Get individual player win statistics')
    async def playerstats(self, interaction: discord.Interaction, player: str, days: int = None):
        try:
            await interaction.response.defer()
            
            if days is not None and days < 0:
                embed = discord.Embed(title="❌ Invalid Input", description="Days must be 0 or greater (0 = last 24 hours, 1 = last 2 days, etc.)", color=0xff0000)
                await interaction.followup.send(embed=embed, ephemeral=True)
                return
            
            player_name = player.strip()
            
            query = {'clan_winners': player_name}
            
            if days is not None:
                from datetime import timedelta
                now = datetime.now(timezone.utc)
                hours_back = (days + 1) * 24
                cutoff_date = now - timedelta(hours=hours_back)
                query['timestamp'] = {'$gte': cutoff_date}
                time_period = "Last 24 Hours" if days == 0 else f"Last {days + 1} Days"
            else:
                time_period = "All Time"
            
            pipeline = [
                {'$match': query},
                {'$facet': {
                    'stats': [{'$group': {
                        '_id': None,
                        'total': {'$sum': 1},
                        'contests': {'$sum': {'$cond': ['$contest', 1, 0]}},
                        'points': {'$sum': {'$cond': ['$contest', {'$multiply': ['$player_count', 2]}, '$player_count']}}
                    }}],
                    'clans': [{'$group': {'_id': '$clan_name', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 3}],
                    'maps': [{'$group': {'_id': '$map', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 3}],
                    'recent': [{'$sort': {'timestamp': -1}}, {'$limit': 5}, {'$project': {'clan_name': 1, 'map': 1}}]
                }}
            ]
            
            result = await db.db.clanwins.aggregate(pipeline).to_list(1)
            
            if not result:
                embed = discord.Embed(
                    title=f"Player Statistics: {player_name}", 
                    description="No wins found for this player.", 
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            data = result[0]
            stats = data.get('stats', [])
            
            if not stats or stats[0].get('total', 0) == 0:
                embed = discord.Embed(
                    title=f"Player Statistics: {player_name}", 
                    description="No wins found for this player.", 
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            stats_data = stats[0]
            total_wins = stats_data.get('total', 0)
            contest_wins = stats_data.get('contests', 0)
            total_points = stats_data.get('points', 0)
            
            top_clans = [(c['_id'], c['count']) for c in data.get('clans', [])]
            top_maps = [(m['_id'], m['count']) for m in data.get('maps', [])]
            recent_wins = data.get('recent', [])
            
            embed = discord.Embed(
                title=f"{player_name} Statistics",
                description=f"{time_period}",
                color=0x2b2d31
            )
            
            stats_text = f"Wins: **{total_wins}** | Contest: **{contest_wins}** ({contest_wins/total_wins*100:.0f}%)\n" if total_wins > 0 else "Wins: **0** | Contest: **0**\n"
            stats_text += f"Points: **{total_points}**"
            embed.add_field(name="Stats", value=stats_text, inline=False)
            
            if top_clans:
                clan_text = "\n".join([f"{i+1}. [{name}] - {count}" for i, (name, count) in enumerate(top_clans)])
                embed.add_field(name="Top Clans", value=clan_text, inline=True)
            
            if top_maps:
                map_text = "\n".join([f"{i+1}. {name} - {count}" for i, (name, count) in enumerate(top_maps)])
                embed.add_field(name="Top Maps", value=map_text, inline=True)
            
            recent_wins = data.get('recent', [])
            if recent_wins:
                recent_text = "\n".join([f"[{win.get('clan_name', 'Unknown')}] {win.get('map', 'Unknown')}" for win in recent_wins])
                embed.add_field(name="Recent", value=recent_text, inline=False)
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"💥 COMMAND ERROR: {e}")
            await interaction.followup.send(embed=discord.Embed(
                title="Error", description="Failed to fetch player statistics.", color=0xff0000
            ))

async def setup(bot):
    await bot.add_cog(PlayerStats(bot))

File: commands/stats/playerstats.js
// https://github.com/viktorexe/terristats-discord-bot
// /playerstats - individual player win statistics with optional day filter.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('playerstats')
    .setDescription('Get individual player win statistics')
    .addStringOption((o) =>
      o.setName('player').setDescription('Player name').setRequired(true),
    )
    .addIntegerOption((o) =>
      o.setName('days').setDescription('Optional day filter').setRequired(false),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const player = interaction.options.getString('player');
      const days = interaction.options.getInteger('days');

      if (days !== null && days < 0) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('❌ Invalid Input')
              .setDescription(
                'Days must be 0 or greater (0 = last 24 hours, 1 = last 2 days, etc.)',
              )
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const playerName = player.trim();
      const query = { clan_winners: playerName };

      let timePeriod;
      if (days !== null) {
        const hoursBack = (days + 1) * 24;
        query.timestamp = { $gte: new Date(Date.now() - hoursBack * 3600 * 1000) };
        timePeriod = days === 0 ? 'Last 24 Hours' : `Last ${days + 1} Days`;
      } else {
        timePeriod = 'All Time';
      }

      const pipeline = [
        { $match: query },
        {
          $facet: {
            stats: [
              {
                $group: {
                  _id: null,
                  total: { $sum: 1 },
                  contests: { $sum: { $cond: ['$contest', 1, 0] } },
                  points: {
                    $sum: {
                      $cond: [
                        '$contest',
                        { $multiply: ['$player_count', 2] },
                        '$player_count',
                      ],
                    },
                  },
                },
              },
            ],
            clans: [
              { $group: { _id: '$clan_name', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 3 },
            ],
            maps: [
              { $group: { _id: '$map', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 3 },
            ],
            recent: [
              { $sort: { timestamp: -1 } },
              { $limit: 5 },
              { $project: { clan_name: 1, map: 1 } },
            ],
          },
        },
      ];

      const result = await db.collection('clanwins').aggregate(pipeline).toArray();
      const data = result[0];
      const stats = data ? data.stats : [];

      if (!data || !stats.length || (stats[0].total || 0) === 0) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`Player Statistics: ${playerName}`)
              .setDescription('No wins found for this player.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const statsData = stats[0];
      const totalWins = statsData.total || 0;
      const contestWins = statsData.contests || 0;
      const totalPoints = statsData.points || 0;

      const topClans = data.clans.map((c) => [c._id, c.count]);
      const topMaps = data.maps.map((m) => [m._id, m.count]);
      const recentWins = data.recent || [];

      const embed = new EmbedBuilder()
        .setTitle(`${playerName} Statistics`)
        .setDescription(timePeriod)
        .setColor(0x2b2d31);

      let statsText =
        totalWins > 0
          ? `Wins: **${totalWins}** | Contest: **${contestWins}** (${((contestWins / totalWins) * 100).toFixed(0)}%)\n`
          : 'Wins: **0** | Contest: **0**\n';
      statsText += `Points: **${totalPoints}**`;
      embed.addFields({ name: 'Stats', value: statsText, inline: false });

      if (topClans.length) {
        embed.addFields({
          name: 'Top Clans',
          value: topClans.map(([n, c], i) => `${i + 1}. [${n}] - ${c}`).join('\n'),
          inline: true,
        });
      }
      if (topMaps.length) {
        embed.addFields({
          name: 'Top Maps',
          value: topMaps.map(([n, c], i) => `${i + 1}. ${n} - ${c}`).join('\n'),
          inline: true,
        });
      }
      if (recentWins.length) {
        embed.addFields({
          name: 'Recent',
          value: recentWins
            .map((w) => `[${w.clan_name || 'Unknown'}] ${w.map || 'Unknown'}`)
            .join('\n'),
          inline: false,
        });
      }

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`COMMAND ERROR: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to fetch player statistics.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
